File: kitchenowl/lib/models/recipe_review.dart
import 'package:kitchenowl/models/model.dart';
import 'package:kitchenowl/models/user.dart';

class RecipeReview extends Model {
  final int? id;
  final int rating;
  final String review;
  final User? user;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  const RecipeReview({
    this.id,
    required this.rating,
    this.review = '',
    this.user,
    this.createdAt,
    this.updatedAt,
  });

  factory RecipeReview.fromJson(Map<String, dynamic> map) {
    DateTime? timestamp(dynamic value) => value is int
        ? DateTime.fromMillisecondsSinceEpoch(value, isUtc: true)
        : value is String
            ? DateTime.tryParse(value)
            : null;

    return RecipeReview(
      id: map['id'],
      rating: map['rating'] ?? 0,
      review: map['review'] ?? '',
      user: map['user'] is Map<String, dynamic>
          ? User.fromJson(map['user'])
          : null,
      createdAt: timestamp(map['created_at']),
      updatedAt: timestamp(map['updated_at']),
    );
  }

  @override
  List<Object?> get props => [id, rating, review, user, createdAt, updatedAt];

  @override
  Map<String, dynamic> toJson() => {
        'rating': rating,
        'review': review,
      };

  @override
  Map<String, dynamic> toJsonWithId() => toJson()..['id'] = id;
}

File: backend/app/models/recipe_review.py
from __future__ import annotations

from datetime import datetime
from typing import TYPE_CHECKING, Any, cast

from sqlalchemy import CheckConstraint, ForeignKey, UniqueConstraint
from sqlalchemy.orm import Mapped

from app import db

Model = db.Model
if TYPE_CHECKING:
    from app.helpers.db_model_base import DbModelBase
    from app.models import Recipe, User

    Model = DbModelBase


class RecipeReview(Model):
    __tablename__ = "recipe_review"
    __table_args__ = (
        UniqueConstraint("recipe_id", "user_id", name="uq_recipe_review_recipe_user"),
        CheckConstraint("rating >= 1 AND rating <= 5", name="ck_recipe_review_rating"),
    )

    id: Mapped[int] = db.Column(db.Integer, primary_key=True)
    recipe_id: Mapped[int] = db.Column(
        db.Integer,
        ForeignKey("recipe.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    user_id: Mapped[int] = db.Column(
        db.Integer,
        ForeignKey("user.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    rating: Mapped[int] = db.Column(db.Integer, nullable=False)
    review: Mapped[str] = db.Column(db.Text(), nullable=False, default="")

    recipe: Mapped["Recipe"] = cast(
        Mapped["Recipe"],
        db.relationship("Recipe", back_populates="reviews", uselist=False),
    )
    user: Mapped["User"] = cast(
        Mapped["User"],
        db.relationship("User", back_populates="recipe_reviews", uselist=False),
    )

    def obj_to_dict(
        self,
        skip_columns: list[str] | None = None,
        include_columns: list[str] | None = None,
    ) -> dict[str, Any]:
        skip = list(skip_columns or [])
        skip.extend(["recipe_id", "user_id", "created_at", "updated_at"])
        res = super().obj_to_dict(skip, include_columns)
        res["user"] = self.user.obj_to_dict() if self.user else None
        res["created_at"] = self._timestamp(self.created_at)
        res["updated_at"] = self._timestamp(self.updated_at)
        return res

    @staticmethod
    def _timestamp(value: datetime | None) -> int | None:
        return round(value.timestamp() * 1000) if value else None
